### ISSUE TITLE
Openpulse signal formalism

### ISSUE THREAD

**Issue by taalexander** [Qiskit/openqasm#126]
This is a discussion topic is for an alternate `signal` representation that could replace the `pulse` representation in #109. There are many ideas in this document, some of which I believe is likely out of scope.

As discussed synchronously here is a proposal for a pulse-level representation based on `signal`s and `channel`s. A `signal` corresponds to the logical signal that must be applied and is mapped to a physical hardware resource which is a `channel`. It is hoped that this proposal is a more "natural" model for pulse-level control and is extendable to varied hardware in the future.

*Note - Some of the syntax/language used within this proposal is different than in #109. I am not tied to any of this, I just found it helped with consistency.*

## Signals

A `signal` is a generalization of a concept of a `pulse`. A signal consists of a discrete time-dependent function `s(t), ℤ->ℂ` of max unit-norm and a clock-value (marker) `t ∈ ℤ, t >= 0`. In a sense, the time `t` is an "index" into the signal. The clock-value is propagated from its child signal with the initial value originating from the clock of the channel on which the signal is applied.

Openpulse should supply many types of builtin signal production operations such as
- `envelope` - A complex envelope - `let signal: envelope = [0.0+1.0j, ..., 1.0+0.0j];`
- `wave` - A signal representing a carrier wave - `let frame: wave = carrier(5e9, 2*pi);`

Signals may be composed via transformation functions to form new signals that are derived from the input parent signals. For example `let new_signal = mix(envelope, wave);` produces a new child signal that is a mixture of its parent envelope and carrier wave signals. Combining signals in this way forms a "signal tree", where signals are edges between signal production/transformation leaves. Within the signal programming model, signals are constructed(composed) from the bottom-up. The root of the tree is the root signal which is eventually supplied to a `channel`. The clock-value of a signal is propagated from its clock. Within this formalism, builtin signal-generation operations such as `gaussian` are `source` nodes that produce a signal (ie., no input edges).
```
    sig1 sig2
      |   |
      v   v
  transformation
        |
        v
       (root) sig3
```
A signal network has a correspondence with a traditional microwave schematic. Unlike qubit operations, signals are not linear and may be reused by splitting them, ie., they may be cloned. In practice, we allow reusing the named edged without requiring an explicit split operation.
```
       sig1
        |
        v
      split
       / \
      v   v
     re  im
      |   |
      v   v
    sig2 sig3
```
A signal is defined by its ancestor signal production/transformation nodes, $s_n(t) = f_{n-1}(s_{n-1}^{(0)}(t), s_{n-1}^{(j)}(t)) = f_{n-1}(f_{n-1}^{(0)}(s_{n-2}^{(0)}, ...), ..., f_{n-2}^{j}(s_{n-2}^{(j)}, ...))$. This automatically gives many nice properties. For example, if all dependent signals are differentiable, so must be the final signal by the chain rule.

When outputting a signal on a `channel` the intermediate signals in the signal network are only useful insofar as to describe the output signal, this enables the compiler to restructure the signal network to better map to the available hardware. For example, if the hardware is capable of providing support for a `carrier`, it may natively represent a signal of the form `mix(envelope, carrier)`, otherwise it may choose to rewrite the signal as a sidebanded envelope.

### Signal production operations
Signal production operations are signal `sources` that maintain clock relationships when coupled to an output channel.
   - `envelope`: Maintain an internal clock that begins at `t=0` for every usage with a channel, eg., $t_{envelope} = t_{ch} - t0_{ch}$. This is to enable the reuse of an envelope signal many times. (*Note- the requirement for this special behavior could be replaced if a normal signal is wrapped in a `set` transformation. This would be beneficial as we could partially use waveforms for example to reuse a section of a gaussian pulse.*). Are parameterized such that `s(0) = 0`, `s(t>=tf) = 0`.
     -  `array` - `let wf1: envelope = [0.1, 1.j, 0.1 - 0.2j, ...];`
     - `gaussian(amp: complex, duration: length, sigma: length)`,
     - `gaussian_square(amp: complex, duration: length, sigma: length, square_width: length)`,
     - `drag(amp: complex, duration: length, sigma: length, beta: float)`,
     - `constant(amp: complex, duration: length)` (aliased by complex scalar? eg., `let const_wf: envelope = 0.5 + 0.5j;`)
   - `wave(amp: complex, freq: float, phase: angle)` - Characterized by `amp`, `freq` and `phase`. This modification occurs at the current time of the signal's local clock. The clock is determined by the clock of the controlling physical channel, eg., $t_{wave} = t_{ch}$. Waves are a named three-tuple and may be modified with dot notation during real-time execution, eg., `wave.phase += pi;`. Updates to a wave's properties will therefore propagate forward in time for future uses as determined by the applied physical channel. In this way, persistent amplitude/phase/frequency-updates may be applied.
     - `carrier` - $Ae^{i(wt+\phi)}$, Sine/Cosine waves can be obtained by deriving with `re`/`im`. Need better name.
     - `sawtooth`
     - `square`
     - `triangle`
   - Opaque signal: An opaque signal could be declared and then matched in the compiler by name `let noise_source: signal = opaque_signal();`. The clock of the opaque signal is controlled by that of the channel it is used with.

### Signal composition operations
Each takes as input one or more signals, applies a transformation, and produces a new signal. Note these do not mutate the input signals but are more akin to piping the signal into a transfer function. Signals are classical and may be copied without a problem. It is up to the compiler to choose how to implement the required transformations on hardware:
 - `shift(sig: signal, time: length)` - Shift the signal by `time` by incrementing the signal clock to be passed to ancestor signal `sig`.
 - `set(sig: signal, time: length)` - Ignore the propagated child signal clock and replace it with `time` to be propagated to all ancestor nodes.
 - `mix(sig: signal, sig: signal) -> signal` - Mix two input signals to produce a new signal. This is equivalent to the product signal `s(t_i) = s_1(t_i) x s_2(t_i)`.
 - `sum(sig: signal, sig: signal) -> signal` - Sum two input signals to produce a new signal.
 - `piecewise(sig0: signal, sig1: signal, time: length) -> signal` - Output `sig
 - `offset(sig: signal, amount: complex) -> signal` - Offset the input signal by an `amount`.
 - `scale(sig: signal, factor: complex) -> signal`(`attenuate`/`amplify`) - Scale the input signal by a `factor`.
 - `conj(sig: signal) -> signal` - Conjugate the input signal.
 - `re(sig: signal) -> signal` - Real component of input signal.
 - `im(sig: signal) -> signal` - Imaginary component of input signal.
 - `abs(sig: signal) -> signal` - Transform signal as norm of input. signal
 - `phase(sig: signal, ang: angle) -> signal` - Signal with relative phase, `ang`.

## Channels
`channel`s are the interface between the logical `signal` world and the physical world. They are representations of the IO ports of the device. Within a pulse program a channel's primary responsibility is to represent a "real-world clock" that can only advance through `delay` instructions, `tx` instructions for `txchannel`s and `rx` for `rxchannel`s as all signal operations are eventually defined with respect to channel clocks.

The base channel type is `channel`. There are two primary channel subtypes:
- `txchannel` - A channel for emitting signals on a physical port.
- `rxchannel` - A channel for capturing signals to produce a logical signal.

Signals are transmitted and received with the `tx` and `rx` functions on the `txchannel`s and `rxchannel`s respectively.

### Channel signal output operations
- `transmit(ch: txchannel, sig: signal, duration: length)`: Transmit the real component of a `signal` on a `txchannel` resource for a `length` of time. Advances the real-time clock of this channel by `duration`.

If channel production operations are signal `sources`, the `transmit` operation is the signal `sink` (ie., no output edges). Piping the signal out of the logical Openpulse domain into the physical hardware.

### Channel signal input operations
- `receive(ch: rxchannel, duration: length) -> signal`: Capture a real `signal` from a `rxchannel` resource for a `length` of time. Increments the target channel's clock by `duration`. The produced signal's clock is shifted to `time(ch)` so as to enable demodulate the signal against the measurement carrier tone.

### Channel operations:
- `time(ch: channel) -> length` - Get current real clock-time of the channel. *Note - this is likely not required*
- `barrier(ch1: channel, ..., chn: channel)` - Advance the real clocks of each input channel to `max(time(ch1), ..., time(chn))`.
- `delay(ch: channel, duration: length)` - Increment the real signal clock by a duration.

### Channel Getters
*Note - Added so that document is self-contained, these are supported in the existing proposal already under different names.*
The following operations are supported to fetch the configured channels for the specific device.
- `txch(q0: qubit, ..., qn: qubit, name: str) -> txchannel`
- `rxch(q0: qubit, ..., qn: qubit, name: str) -> rxchannel`

Alternatively, one could declare a channel of a given name directly and rely on the compiler to bind this to a physical channel on the device by the declared name.
```qasm
let d0: txchannel;
```

### Cross-resonance like gate
Playing simultaneous pulses on two separate channels with a shared phase/frequency relationship. Similar to requirements for cross-resonance gate.
```qasm
// Initialize
let d0 = txch(0, "drive");
let d1 = txch(1, "drive");
let carrier = exp(1.0, 5e9, 0);
// Do a bunch of operations incrementing the channels times.
// Synchronize clocks.
barrier(d0, d1);
let wf2 = gaussian_square(1., 1024dt, 32dt, 128dt);;
let wf2 = gaussian_square(0.1, 1024dt, 32dt, 128dt);
// Produce new carrier with phase shifted from derived
let offset_carrier = phase(carrier, pi/2);
tx(d0, mix(wf1, carrier), 1024dt);
tx(d1, mix(wf2, pi2_offset_carrier), 1024dt);
```

### Measuring a qubit.
Here we use a kernel operation on the signal to discriminate the signal to a bit.
```qasm
kernel discriminate(sig: signal, len: length, time: length) -> bit;
let m0 = txch(0, "measure");
let cap0 = rxch(0, "capture");

// Force time of carrier to 0 for consistent phase for discrimination.
let carrier = set(exp(1.0, 5e9, 0), 0);
barrier(q0, cap0);
tx(m0, mix(carrier, gaussian_square(1.0, 16000dt, 262dt, 13952dt)));
// Measure and demodulate measured signal.
let output = mix(rx(cap0, 16000dt), phase(carrier, pi));
// Pass signal to kernel to be discriminated.
let result: bit = discriminate(output, 16000dt);
```

### Clocking example
```qasm
let d0: txchannel; // t=0
let d1: txchannel; // t=0

let env0: envelope = [0.0, 1.0, 0.0];
let carrier: wave = carrier(1.0, 5e9, 0.0);

let sig0: signal = mix(env0, carrier);
//        env0[t=0] carrier[t=0]
//             \     /
//                |
//              mix[t=0]
//                |
//              sig0[t=0]

let sig1: signal = shift(sig0, 10);
//        env0[t=0] carrier[t=0]
//             \     /
//                |
//              mix[0]
//                |
//            shift(10)[0]
//                |
//              sig1[t=0]

// Advance physical channel clock by transmitting instruction
tx(d0, sig0, 3); // t=0->3
//    env0[t=0->3] carrier[t=0->3]
//             \     /
//                |
//            mix[t=0->3]
//                |
//            sig0[t=0->3]

// Envelope has clock forced to t=0
tx(d0, sig0, 3); // t=3->6
//    env0[t=0->3] carrier[t=3->6]
//             \     /
//                |
//            mix[t=3->6]
//                |
//            sig0[t=3->6]

// Inheritance of clock shift in signal
tx(d0, sig1, 10); // t=3->6
//    env0[t=0->10] carrier[t=16->26]
//             \     /
//                |
//            mix[t=16->26]
//                |
//            shift(10)[t=16->26]
//                |
//              sig1[t=6->16]

// Channel d1's clock has not yet advanced.
// This enables scheduling in parallel across channels.
tx(d1, sig0, 3); // t=0->3
//    env0[t=0->3] carrier[t=0->3]
//             \     /
//                |
//            mix[t=0->3]
//                |
//            sig0[t=0->3]
```

## Questions
- Decide how we wish to formulate clocking language. Is it in terms of a "global clock" and a "signal clock", or a "global clock" and a "signal marker".
- Do we consider supporting non-complex pulses domains. For example, it may be valuable to operate in a domain with parameterized integer or floating-point precision using `int[size]`/`float[size]` to specify a signal faithfully for target hardware.
- Do we use special functions for modifying/composing signals such as `invert`, `scale`, `mix`, `couple` or do we support infix operators like `~`, `signal * 0.2`, `signal1 ⊗ signal2`, `signal1 + signal2`, respectively. *note - we could unify `*` and `⊗` by aliasing constant signals as floats/complex values. These could also be used to modify `envelope`/`pulse` parameters with dot notation, eg., `carrier.phase += pi`. There is a difference between producing a new signal and modifying an existing one.
- Where possible signals should likely be parameterized with lengths to enable more expressive signal descriptions. Naturally, we might also want to express `frequency` as a length, should this be in terms of inverse length or just a period?
- Assembly or C-style function-call syntax?

*Note - [faust](https://faust.grame.fr/) is a functional programming language for DSP that has many interesting relevant concepts.*

## Future Extensions
- Signal filtering, eg., a FIR filter could look like `output = fir(input, b0, b1, b2, ...)`.
- How do we handle multiplexing measurement?
- How do we define global signals, or at the top level?
-

**Comment by zachschoenfeld33** [comment]
I like the proposal. Some minor comments:
1. Is `mix` multiplication? I don't think this is ever clarified, but that's what it seems to be.
2. The `tx/rx` functions may be a bit unclear compared to `transmit/receive`. There is also the issue of overlap with x rotation gate (ie `qc.rx(qubit)`).
3. Regarding the special functions question: Could we use operators and define equivalent special functions as in typical classical languages. For instance, how `+` is equivalent `__add(sig1, sig2)__` in Python? I think you would produce a new signal, but you can choose whether or not to assign it (like standard integer or string operators).

**Comment by taalexander** [comment]
1. Yes, open to other names or binary operators (see the end). Within a DSL such as openpulse it may make sense to have the domain term `mix` be used.
2. Fair enough, although within a DSL I expect this confusion to be short-term, and may make up for having to repetitively type `transmit`.
3. This works within the data model of Python but that is significantly more complex than OpenQASM. I would recommend just defining a binary operator equivalent of `mix` as `add` directly, but would maybe consider this as a follow-on feature at a later point in time. Having a separation between signals and normal scalar values can be an important distinction.

**Comment by zachschoenfeld33** [comment]
Could the definition of `mix` add clarification that it multiplies? Current definition is 
`mix(sig: signal, sig: signal) -> signal` - Mix two input signals to produce a new signal.

**Comment by taalexander** [comment]
Explore Praschandt's suggestion to send multiple things to TX/RX.

**Comment by stevenheidel** [comment]
> piecewise(sig0: signal, sig1: signal, time: length) -> signal - Output `sig

Looks like the sentence got cut off. I'm assuming it should say something like output sig0 followed by sig1

**Comment by psivaraj** [comment]
I like the way this is forcing me to think about things, so kudos! Just a couple of comment/questions for now: 

1. The concept of a `channel` and `signal` feel useful to me -- the latter actually just holding a frequency, phase, and time. Like you said, the power is that a `signal` can be shared between `channel`s to signify a true physics relationship, and the time advancement of a `channel` and `signal` seem straightforward during a `transmit/receive` -- both advance for the time requested. 

    But I'm not seeing the value of making the `envelope` a `signal` type. Why can't it be a separate type that a `transmit`/`receive` ingest? e.g. 

        transmit(ch: txchannel, sig: signal, env: envelope, duration: length)

    The operation would be to perform complex multiplication between `signal` and `envelope` with the side-effect of advancing the time of `channel` and `signal`. You're basically there in your "Cross-resonance like gate" example, and I don't see where it's useful for an `envelope` to have a persistent time as it seems to just reset in your "Clocking example".  

2. I'm also little unclear on the idea

      > Within the signal programming model, signals are constructed(composed) from the bottom-up. The root of the tree is the root signal which is eventually supplied to a channel. The clock-value of a signal is propagated from its clock.

    Maybe I missed it or am misunderstanding, but  I'm nervous about the complexity. Why can't  the `signal` you're inputting into a `transmit/receive/shift` be the one and only `signal` that is advanced? Could you describe an example where this tree structure would be useful?

**Comment by eggerdj** [comment]
I like what I am reading. Conceptually I often think of a signal as `envelop(t)*exp[i (w t+ phase)]`. So this makes sense to me.

1. Do you intend to be able to support hardware in which the output devices have different sampling rates? (which I'm guessing would be tricky).
2. What is the purpose of `shift`? In which case is this needed?
3. what does `piecewise` do?
4. You defined `phase` which as far as I understand adds a phase to the signal. Why not define a freq as well?

**Comment by caryan** [comment]
I definitely echo some of the comments that giving the user more of the DSP to hold and manipulate in their hands feels powerful. The one thing I kept getting caught up by though is reasoning about the timing of the in-place updates such as phase shifts or scaling e.g. `carrier.phase += pi/2`. By having declarative signals and by separating the timekeeper (now the `channel`) from the in-place frame actions (now the `wave`) it feels actions become non-local and harder to reason about particularly when sharing a carrier between channels. E.g. taking something similar to the clocking example:

```
let d0: txchannel; // t=0
let d1: txchannel; // t=0

let env0: envelope = [0.0, 1.0, 0.0];
let carrier: wave = carrier(1.0, 5e9, 0.0);

// play a pulse on channel d0
tx(d0, mix(env0, carrier))
// frame update but what time does this frame update occur?
carrier.phase += pi/2
// play a pulse on channel d1
// what phase does the carrier have for this pulse?
tx(d1, mix(env0, carrier))
```

Something similar happens with the signal tree and in-place operators. This issue would go away if we dropped in-place operations and made a copy every time but that solution seems to make the most common operation I'll want to do (frame shifts) annoying to have to generate to generate a new label every time.

A few other minor bits and bobs:
* `piecewise` seems to be for concatenation but then why does it take a `length`?
* signals have an intrinsic length so why does transmit also have a `length` and what happens if they conflict?
* is the sample point defined for array type signals? Related, "Are parameterized such that s(0) = 0, s(t>=tf) = 0" seems to force the endpoints to be zero.

**Comment by taalexander** [comment]
@caryan 
> Something similar happens with the signal tree and in-place operators. This issue would go away if we dropped in-place operations and made a copy every time but that solution seems to make the most common operation I'll want to do (frame shifts) annoying to have to generate to generate a new label every time.

This is exactly the path I went down as well and I echo your sentiments about the in-place updates. I'll try and put my thinking cap on and resolve this.

**Comment by taalexander** [comment]
@psivaraj 
> I like the way this is forcing me to think about things, so kudos! Just a couple of comment/questions for now:
> 
> 1. The concept of a `channel` and `signal` feel useful to me -- the latter actually just being a carrier that holds a frequency, phase, and time. Like you said, the power is that a `signal` can be shared between `channel`s to signify a true physics relationship, and the time advancement of a `channel` and `signal` seem straightforward during a `transmit/receive` -- both advance for the time requested.
>    But I'm not seeing the value of making the `envelope` a `signal` type. Why can't it be a separate type that a `transmit`/`receive` ingest? e.g.
>    ```
>     transmit(ch: txchannel, sig: signal, env: envelope, duration: length)
>    ```
>    
>    
>    The operation would be to perform complex multiplication between `signal` and `envelope` with the side-effect of advancing the time of `channel` and `signal`. You're basically there in your "Cross-resonance like gate" example, and I don't see where it's useful for an `envelope` to have a persistent time as it seems to just reset in your "Clocking example".

I think as you say "The operation would be to perform complex multiplication between `signal` and `envelope` with the side-effect of advancing the time of `channel` and `signal`". It seems like special-casing a common case which may be fine. Perhaps a more gracefully solution would be to allow the equivalent of

`transmit(ch, duration, *signals)` so that `transmit(ch, duration, envelope, carrier)` and ``transmit(ch, duration, mix(evelope, carrier))` would be equivalent.


> 2. I'm also little unclear on the idea
>    > Within the signal programming model, signals are constructed(composed) from the bottom-up. The root of the tree is the root signal which is eventually supplied to a channel. The clock-value of a signal is propagated from its clock.
>    
>    
>    Maybe I missed it or am misunderstanding, but  I'm nervous about the complexity. Why can't  the `signal` you're inputting into a `transmit/receive/shift` be the one and only `signal` that is advanced? Could you describe an example where this tree structure would be useful?

Perhaps the wording is unclear. Since the input signal is composed of other signals the clock value must be propagated to these "signal functions" to obtain their output value at the current top-level time. Pythonically this might be equivalent to the call tree below:

```python
def s3(t):
    return s2(t)*s1(t)
def s2(t):
    return math.cos(w2*t)
def s1(t):
    # shifted signal
    return s0(t+10)
def s0(t):
    return math.cos(w0*t)
```

When `transmit(ch, s3, duration)` is called `ch[t_curr]`, will emit `s3(t_curr), ..., s3(t_curr+duration-1)`.

**Comment by taalexander** [comment]
@eggerdj 
> I like what I am reading. Conceptually I often think of a signal as `envelop(t)*exp[i (w t+ phase)]`. So this makes sense to me.
> 
1. Do you intend to be able to support hardware in which the output devices have different sampling rates? (which I'm guessing would be tricky).

Absolutely. The model above could be extended to the continuous domain such that `t ∈ R`, where there are no problems (besides that it doesn't represent the reality of signal generation). One of two strategies can be taken. Set `dt` to be the least common multiple of all sampling rates. This is the most simplistic approach but limits resolution. The other approach is to set `dt` to be some common factor of all relevant sampling rates. It would be up to the hardware to error at compilation if the desired signal could not be resolved (eg. some sample multiple like the restriction of 16 in our current systems). This is assuming the system is engineered sanely and this is feasible. Even if it is not there is likely some tolerance around timing alignment and this could be expressed with an attribute on a barrier `barrier<"epsilon"=1e-10>(ch1, ch2)`. 


> 2. What is the purpose of `shift`? In which case is this needed?

The idea here is that it can enable the construction of more complicated piecewise signals, eg., a gaussian square I could define one `gaussian` pulse for the rise/fall and then shift it appropriately to sample the rise/fall components, combined with piecewise of a `constant` pulse to build the final signal.

> 3. what does `piecewise` do?

It combines two signals to produce a new signal in which the `sig1, t<= t_in`, `sig2, t>t_in`. See the example above.

> 4. You defined `phase` which as far as I understand adds a phase to the signal. Why not define a freq as well?

A phase is just equivalent to multiplication by a complex factor for all signal outputs, this might be better named as `angle`? I think its not straightforward what `freq` would be in this context.

**Comment by zachschoenfeld33** [comment]
There are two clocks and I think this could cause some confusion. The signal itself has a clock, which defines the `t_i` for the first pulse sample. There is also the clock for each channel, which defines where pulses are played relative to each other. It might help to clarify this somewhat. I do agree both may be useful.

**Comment by taalexander** [comment]
There aren't two clocks. The only clock exists on the channel. A signal is merely a function accepting a clock value from a channel to produce a value `s(t), ℤ->ℂ` as defined above.

**Comment by zachschoenfeld33** [comment]
Apologies. What I meant is that one might confuse the time index for the signal itself, that is `t_i` for `s(t_i)`, with the time which represents the clock value for a given channel.

The description may lead to this confusion. You mention "signal's local clock" and the "target channel's clock". This makes is seem like there are two clocks. Instead, I think we should distinguish between the signal start index (or start time index) and the clock time on the channels. 

It might even be easier to just think about a signal as a discrete set of samples, rather than possibly shifted or modified from some prior signal. 

For instance, if you have 
```
let env0: envelope = [0.0, 1.0, 0.0, 1.0, 1.0];
let sig0: signal = shift(env0, 2);
```
then `sig0` is just the array `[0.0, 1.0, 1.0]`. 

It may be more confusing to think of it as shifted in time by two samples from `env0`.

**Comment by taalexander** [comment]
After thinking on our discussion regarding mutating signals last week, I am not so sure a timestamp is required to mutate a signal. The timestamp is inherited the next time the signal is used. There are two concepts of order in the signal formalism.

1. Instruction ordering. This is how we synchronize instructions on channels and signals. Instruction execution is based on the order of the instructions in the program but does not imply a modification of the clock-value directly, these are determined by actions on channels.
2. Time ordering. Transmission instructions increment the time-value. This time value gets *input* from the 

The only thing that matters from the perspective of updating a signal when played is the definition of said signal and the time-value it receives. An update to the signal is defined by its ordering, this is independent of the time it is made at.

<examples to follow>

**Comment by zachschoenfeld33** [comment]
I think some additional clarification for the `rxchannel` could be beneficial. We should note that the output is a raw signal to which one can apply qasm3 kernel functions to get IQ data (what in today's open pulse is called `kernel`) and binary data (what in today's open pulse in called `discriminate`).

Additionally, one potential benefit of the `rxchannel` is it can track timing of data acquisition. But how can one delay on this channel? One option is barriers. But we may want to allow for some notion of adding a delay on an acquire channel without actually capturing a signal for this time.

**Comment by zachschoenfeld33** [comment]
Closing as this proposal is addressed in https://github.com/Qiskit/openqasm/pull/184.
